### PR TITLE
Add listCollateralTokens rpc

### PR DESCRIPTION
**Pull request by jingyi2811**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Add listCollateralTokens rpc

#### Which issue(s) does this PR fixes?:

Fixes part of #488

### CODE REVIEW

File: packages/jellyfish-api-core/__tests__/category/loan/listCollateralTokens.test.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      token: 'TSLA',
+      factor: new BigNumber(1),
+      priceFeedId: priceFeedId2,
+      activateAfterBlock: 170
+    }])
+    await testing.generate(1)
+
+    const data = await testing.rpc.loan.listCollateralTokens()
+    expect(data).toStrictEqual({
+      [collateralTokenId1]: { activateAfterBlock: 160, factor: 0.5, priceFeedId: priceFeedId1, token: 'AAPL' },
```
Comment (Jouzo) [inline packages/jellyfish-api-core/__tests__/category/loan/listCollateralTokens.test.ts:58]: `factor` should be BigNumber.
I don't know how this should be done as it is nested and not handled by `{ factor: 'bignumber' }`
cc @fuxingloh

Comment (fuxingloh) [reply packages/jellyfish-api-core/__tests__/category/loan/listCollateralTokens.test.ts:58]: 2 Problems

1. `{ factor: 'bignumber' }` only works for list and object, not for a map with keys. Difficult to do a full mapping, we should just resort to mapping everything as BigNumber
2. `C++` shouldn't return a `Map` for a pagination/list query, counter-intuitive. It losses the natural ordering and is semantically not a sound way to deal with list/cursor operations. They should return an array.

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 2]
```diff
@@ -131,6 +140,17 @@ export interface SetCollateralToken {
...
 
+export interface CollateralTokensData {
+  [key: string]: CollateralTokenDetail
+}
+
+export interface CollateralTokenDetail {
+  token: string
+  factor: BigNumber
+  priceFeedId: string
+  activateAfterBlock: BigNumber
```
Comment (surangap) [inline packages/jellyfish-api-core/src/category/loan.ts:151]: optional. https://github.com/DeFiCh/ain/blob/a43fda7d0842ff1c90ab39880df38077ba36bd91/src/masternodes/rpc_loan.cpp#L15

File: packages/jellyfish-api-core/__tests__/category/loan/listCollateralTokens.test.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      token: 'TSLA',
+      factor: new BigNumber(1),
+      priceFeedId: priceFeedId2,
+      activateAfterBlock: 130
+    }])
+    await testing.generate(1)
+
+    const data = await testing.rpc.loan.listCollateralTokens()
+    expect(data).toStrictEqual({
+      [collateralTokenId1]: { activateAfterBlock: new BigNumber(120), factor: new BigNumber(0.5), priceFeedId: priceFeedId1, token: 'AAPL' },
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/listCollateralTokens.test.ts:58]: Also, add a collateral token without `activateAfterBlock` to the test.

File: packages/jellyfish-api-core/src/category/loan.ts (hunk 2 above)
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:151]: For setCollateralToken, yes, activateAfterBlock is optional.  However, listCollateralTokens always return activateAfterBlock value. If you don't pass in activateAfterBlock in the setCollateralToken, you will retrieve the current block when the setCollateralToken executed.

Comment (surangap) [reply packages/jellyfish-api-core/src/category/loan.ts:151]: yeah true. https://github.com/DeFiCh/ain/blob/13cc9108141fdf4ed0bf93adf8b4b61fd516ebd0/src/masternodes/mn_checks.cpp#L1830
ignore.

**Review by surangap — DISMISSED** [review]

File: packages/jellyfish-api-core/src/category/loan.ts (hunk 2 above)
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:151]: Anyway I improve the test I wrote earlier

**Review by Jouzo — APPROVED** [review]

**Review by surangap — APPROVED** [review]

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 4]
```diff
@@ -131,6 +140,17 @@ export interface SetCollateralToken {
   activateAfterBlock?: number
 }
 
+export interface CollateralTokensData {
+  [key: string]: CollateralTokenDetail
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/loan.ts:144]: node side should probably change this to array

cc @surangap

**Review by fuxingloh — APPROVED** [review]

File: packages/jellyfish-api-core/src/category/loan.ts (hunk 4 above)
Comment (Jouzo) [reply packages/jellyfish-api-core/src/category/loan.ts:144]: Diego is currently working on changing list* to return arrays

**Pull request merged**